### PR TITLE
Add DNS configuration and leak tests

### PR DESCRIPTION
**Pull request by dlon**
These tests ensure that there are no DNS leaks, both inside and outside of the tunnel, for the default resolver and for custom DNS. They also make sure that outbound traffic *is* observed when expected.

In addition to leak tests, the PR also includes tests checking whether the system uses the correct resolver:

* `test_dns_config_default`
* `test_dns_config_custom_private`
* `test_dns_config_custom_public`
* `test_content_blockers`

Fixes DES-56.
Fixes DES-55.

### CODE REVIEW

**Review by Jontified — CHANGES_REQUESTED** [review]

File: scripts/setup-network.sh [hunk 1]
```diff
@@ -30,6 +30,38 @@ ip addr add dev lan-mullvadtest 172.29.1.200
...
+AllowedIPs = 192.168.15.2
+
+CONF
+
+trap "rm $temp_wg_conf" EXIT TERM
+
+ip link add dev wg-relay0 type wireguard
+ip addr add dev wg-relay0 192.168.15.1 peer 192.168.15.2
+wg setconf wg-relay0 ${temp_wg_conf}
+ip link set up dev wg-relay0
```
[the remark this answers is not in the document]
Comment (dlon) [reply scripts/setup-network.sh:63]: It should probably be destroyed in `destroy-network.sh`. It should also be destroyed automagically when the network namespace is destroyed.

File: scripts/setup-network.sh [hunk 2]
```diff
@@ -30,6 +30,38 @@ ip addr add dev lan-mullvadtest 172.29.1.200
...
+PrivateKey = gLvQuyqazziyf+pUCAFUgTnWIwn6fPE5MOReOqPEGHU=
+ListenPort = 51820
+
+[Peer]
+PublicKey = h6elqt3dfamtS/p9jxJ8bIYs8UW9YHfTFhvx0fabTFo=
+AllowedIPs = 192.168.15.2
+
+CONF
+
+trap "rm $temp_wg_conf" EXIT TERM
```
[the remark this answers is not in the document]
Comment (dlon) [reply scripts/setup-network.sh]: Good catch! Fixed.

File: scripts/setup-network.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (dlon) [reply scripts/setup-network.sh:63]: I've added it to the cleanup script.

File: test-manager/src/tests/dns.rs [hunk 3]
```diff
@@ -0,0 +1,566 @@
...
+    let peer_addr: SocketAddr = "172.29.1.200:51820".parse().unwrap();
+
+    let tun_self_addr: Ipv4Addr = Ipv4Addr::new(192, 168, 15, 2);
+    let tun_peer_addr: Ipv4Addr = Ipv4Addr::new(192, 168, 15, 1);
+
+    let public_key =
+        wireguard::PublicKey::from_base64("7svBwGBefP7KVmH/yes+pZCfO6uSOYeGieYYa1+kZ0E=").unwrap();
+    let private_key = wireguard::PrivateKey::from(
+        TryInto::<[u8; 32]>::try_into(
+            base64::decode("mPue6Xt0pdz4NRAhfQSp/SLKo7kV7DW+2zvBq0N9iUI=").unwrap(),
```
[the remark this answers is not in the document]
Comment (dlon) [reply test-manager/src/tests/dns.rs]: Fixed.

File: test-manager/src/tests/dns.rs [hunk 4]
```diff
@@ -0,0 +1,666 @@
...
+            0,
+            "expected no tunnel packets on port 53"
+        );
+
+        //
+        // Examine non-tunnel traffic
+        //
+
+        assert!(
+            non_tunnel_result.packets.len() >= 2,
```
Comment (Jontified) [inline test-manager/src/tests/dns.rs:307]: Nit: why not have `==`?

File: test-manager/src/tests/dns.rs [hunk 5]
```diff
@@ -0,0 +1,566 @@
...
+use super::helpers::update_relay_settings;
+
+/// How long to wait for expected "DNS queries" to appear
+const MONITOR_TIMEOUT: Duration = Duration::from_secs(5);
+
+/// Test whether DNS leaks can be produced when using the default resolver. It does this by
+/// connecting to a custom WireGuard relay on localhost and monitoring outbound DNS traffic in (and
+/// outside of) the tunnel interface.
+///
+/// The test succeeds if and only if all expected outbound packets on port 53 are observed.
```
[the remark this answers is not in the document]
Comment (dlon) [reply test-manager/src/tests/dns.rs]: I've clarified this.

File: test-manager/src/tests/dns.rs [hunk 6]
```diff
@@ -0,0 +1,666 @@
...
+    MonitorOptions,
+};
+
+use super::helpers::update_relay_settings;
+
+/// How long to wait for expected "DNS queries" to appear
+const MONITOR_TIMEOUT: Duration = Duration::from_secs(5);
+
+const TUN_LOCAL_ADDR: Ipv4Addr = Ipv4Addr::new(192, 168, 15, 2);
+const TUN_REMOTE_ADDR: Ipv4Addr = Ipv4Addr::new(192, 168, 15, 1);
```
Comment (Jontified) [inline test-manager/src/tests/dns.rs]: Perhaps we should reference that this is set in the `setup_network.sh` script?

File: test-manager/src/tests/dns.rs [hunk 7]
```diff
@@ -0,0 +1,666 @@
...
+        .expect("failed to obtain tunnel IP");
+
+    log::debug!("Tunnel (guest) IP: {tunnel_ip}");
+    log::debug!("Non-tunnel (guest) IP: {guest_ip}");
+
+    let monitor = create_monitor();
+
+    let _ = rpc.resolve_hostname("mullvad.net".to_owned()).await;
+
+    let result = monitor.wait().await.unwrap();
```
Comment (Jontified) [inline test-manager/src/tests/dns.rs:584]: nit: `monitor_results`?

**Review by Jontified — CHANGES_REQUESTED** [review]

**Review by Jontified — APPROVED** [review]

File: test-manager/src/tests/dns.rs (hunk 4 above)
Comment (dlon) [reply test-manager/src/tests/dns.rs:309]: Answered previously.

File: test-manager/src/tests/dns.rs [hunk 8]
```diff
@@ -0,0 +1,680 @@
...
+                pkt.destination, whitelisted_dest,
+                "unexpected non-tunnel packet on port 53"
+            );
+        }
+    }
+
+    Ok(())
+}
+
+/// Test whether the expected default DNS resolver is used by `gethostbyaddr` (via `ToSocketAddrs`).
```
Comment (dlon) [inline test-manager/src/tests/dns.rs:324]: Should say: `getaddrinfo`

**Review by Jontified — APPROVED** [review]

**Pull request merged**